File: js/packages/teams-ai/src/authentication/OAuthBotAuthentication.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import {
    DialogContext,
    DialogSet,
    DialogState,
    DialogTurnResult,
    DialogTurnStatus,
    OAuthPrompt
} from 'botbuilder-dialogs';
import {
    Storage,
    TeamsSSOTokenExchangeMiddleware,
    TurnContext,
    TokenResponse,
    CardFactory,
    Attachment
} from 'botbuilder';
import { BotAuthenticationBase } from './BotAuthenticationBase';
import { Application } from '../Application';
import { TurnState } from '../TurnState';
import { TurnStateProperty } from '../TurnStateProperty';
import { OAuthSettings } from './Authentication';
import { getSignInResource } from './UserTokenAccess';

/**
 * @internal
 *
 * Handles authentication for Teams bots.
 * @template TState - The type of the turn state object.
 */
export class OAuthBotAuthentication<TState extends TurnState> extends BotAuthenticationBase<TState> {
    private _oauthPrompt: OAuthPrompt;
    private _oauthSettings: OAuthSettings;

    /**
     * Initializes a new instance of the OAuthBotAuthentication class.
     * @param {Application} app - The application object.
     * @param {OAuthSettings} oauthSettings - The settings for OAuthPrompt.
     * @param {string} settingName - The name of the setting.
     * @param {Storage} storage - The storage object for storing state.
     */
    public constructor(
        app: Application<TState>,
        oauthSettings: OAuthSettings, // Child classes will have different types for this
        settingName: string,
        storage?: Storage
    ) {
        super(app, settingName, storage);

        // Create OAuthPrompt
        this._oauthPrompt = new OAuthPrompt('OAuthPrompt', oauthSettings);
        this._oauthSettings = oauthSettings;

        // Handles deduplication of token exchange event when using SSO with Bot Authentication
        app.adapter.use(new FilteredTeamsSSOTokenExchangeMiddleware(this._storage, oauthSettings.connectionName));
    }

    /**
     * Run or continue the OAuthPrompt dialog and returns the result.
     * @param {TurnContext} context - The turn context object.
     * @param {TState} state - The turn state object.
     * @param {string} dialogStateProperty - The name of the dialog state property.
     * @returns {Promise<DialogTurnResult<TokenResponse>>} A promise that resolves to the dialog turn result containing the token response.
     */
    public async runDialog(
        context: TurnContext,
        state: TState,
        dialogStateProperty: string
    ): Promise<DialogTurnResult<TokenResponse>> {
        const dialogContext = await this.createDialogContext(context, state, dialogStateProperty);
        let results = await dialogContext.continueDialog();
        if (results.status === DialogTurnStatus.empty) {
            const card = await this.createOAuthCard(context);
            results = await dialogContext.beginDialog(this._oauthPrompt.id, {
                prompt: {
                    attachments: [card]
                }
            });
        }
        return results;
    }

    /**
     * Continue the OAuthPrompt dialog and returns the result.
     * @param {TurnContext} context - The turn context object.
     * @param {TState} state - The turn state object.
     * @param {string} dialogStateProperty - The name of the dialog state property.
     * @returns {Promise<DialogTurnResult<TokenResponse>>} A promise that resolves to the dialog turn result containing the token response.
     */
    public async continueDialog(
        context: TurnContext,
        state: TState,
        dialogStateProperty: string
    ): Promise<DialogTurnResult<TokenResponse>> {
        const dialogContext = await this.createDialogContext(context, state, dialogStateProperty);
        return await dialogContext.continueDialog();
    }

    /**
     * Creates a new DialogContext for OAuthPrompt.
     * @param {TurnContext} context - The turn context object.
     * @param {TState} state - The turn state object.
     * @param {string} dialogStateProperty - The name of the dialog state property.
     * @returns {Promise<DialogContext>} A promise that resolves to the dialog context.
     */
    private async createDialogContext(
        context: TurnContext,
        state: TState,
        dialogStateProperty: string
    ): Promise<DialogContext> {
        const accessor = new TurnStateProperty<DialogState>(state, 'conversation', dialogStateProperty);
        const dialogSet = new DialogSet(accessor);
        dialogSet.add(this._oauthPrompt);
        return await dialogSet.createContext(context);
    }

    /**
     * Creates an OAuthCard that will be sent to the user.
     * @param {TurnContext} context - The turn context.
     * @returns {Promise<Attachment>} The OAuthCard.
     */
    private async createOAuthCard(context: TurnContext): Promise<Attachment> {
        const signInResource = await getSignInResource(context, this._oauthSettings);
        let link = signInResource.signInLink;
        let tokenExchangeResource;

        if (this._oauthSettings.enableSso == true) {
            tokenExchangeResource = signInResource.tokenExchangeResource;
            link = undefined;
        }

        return CardFactory.oauthCard(
            this._oauthSettings.connectionName,
            this._oauthSettings.title,
            this._oauthSettings.text,
            link,
            tokenExchangeResource,
            signInResource.tokenPostResource
        );
    }
}

/**
 * @internal
 * SSO Token Exchange Middleware for Teams that filters based on the connection name.
 */
export class FilteredTeamsSSOTokenExchangeMiddleware extends TeamsSSOTokenExchangeMiddleware {
    private readonly _oauthConnectionName: string;

    public constructor(storage: Storage, oauthConnectionName: string) {
        super(storage, oauthConnectionName);
        this._oauthConnectionName = oauthConnectionName;
    }

    public async onTurn(context: TurnContext, next: () => Promise<void>): Promise<void> {
        // If connection name matches then continue to the Teams SSO Token Exchange Middleware.
        if (context.activity.value?.connectionName == this._oauthConnectionName) {
            await super.onTurn(context, next);
        } else {
            await next();
        }
    }
}

File: dotnet/packages/Microsoft.TeamsAI/Microsoft.TeamsAI/Application/Authentication/Bot/OAuthBotAuthentication.cs
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Dialogs;
using Microsoft.Bot.Schema;
using Microsoft.Teams.AI.Application.Authentication.Bot;
using Microsoft.Teams.AI.State;

namespace Microsoft.Teams.AI
{
    /// <summary>
    /// Handles authentication for bot in Teams using OAuth Connection.
    /// </summary>
    internal class OAuthBotAuthentication<TState> : BotAuthenticationBase<TState>
        where TState : TurnState, new()
    {
        private readonly OAuthPrompt _oauthPrompt;
        private readonly OAuthSettings _oauthSettings;

        /// <summary>
        /// Initializes the class
        /// </summary>
        /// <param name="app">The application instance</param>
        /// <param name="oauthSettings">The OAuth prompt settings</param>
        /// <param name="settingName">The name of current authentication handler</param>
        /// <param name="storage">The storage to save turn state</param>
        public OAuthBotAuthentication(Application<TState> app, OAuthSettings oauthSettings, string settingName, IStorage? storage = null) : base(app, settingName, storage)
        {
            this._oauthSettings = oauthSettings;

            // Create OAuthPrompt
            this._oauthPrompt = new OAuthPrompt("OAuthPrompt", this._oauthSettings);

            // Handles deduplication of token exchange event when using SSO with Bot Authentication
            app.Adapter.Use(new FilteredTeamsSSOTokenExchangeMiddleware(storage ?? new MemoryStorage(), settingName));
        }

        /// <summary>
        /// Continue the authentication dialog.
        /// </summary>
        /// <param name="context">The turn context</param>
        /// <param name="state">The turn state</param>
        /// <param name="dialogStateProperty">The property name for storing dialog state.</param>
        /// <param name="cancellationToken">The cancellation token</param>
        /// <returns>Dialog turn result that contains token if sign in success</returns>
        public override async Task<DialogTurnResult> ContinueDialog(ITurnContext context, TState state, string dialogStateProperty, CancellationToken cancellationToken = default)
        {
            DialogContext dialogContext = await this.CreateDialogContextAsync(context, state, dialogStateProperty, cancellationToken);
            return await dialogContext.ContinueDialogAsync(cancellationToken);
        }

        /// <summary>
        /// Run or continue the authentication dialog.
        /// </summary>
        /// <param name="context">The turn context</param>
        /// <param name="state">The turn state</param>
        /// <param name="dialogStateProperty">The property name for storing dialog state.</param>
        /// <param name="cancellationToken">The cancellation token</param>
        /// <returns>Dialog turn result that contains token if sign in success</returns>
        public override async Task<DialogTurnResult> RunDialog(ITurnContext context, TState state, string dialogStateProperty, CancellationToken cancellationToken = default)
        {
            DialogContext dialogContext = await this.CreateDialogContextAsync(context, state, dialogStateProperty, cancellationToken);
            DialogTurnResult results = await dialogContext.ContinueDialogAsync(cancellationToken);
            if (results.Status == DialogTurnStatus.Empty)
            {
                Attachment card = await this.CreateOAuthCard(context, cancellationToken);
                Activity messageActivity = (Activity)MessageFactory.Attachment(card);
                PromptOptions options = new()
                {
                    Prompt = messageActivity,
                };

                results = await dialogContext.BeginDialogAsync(this._oauthPrompt.Id, options, cancellationToken);
            }
            return results;
        }

        private async Task<DialogContext> CreateDialogContextAsync(ITurnContext context, TState state, string dialogStateProperty, CancellationToken cancellationToken = default)
        {
            IStatePropertyAccessor<DialogState> accessor = new TurnStateProperty<DialogState>(state, "conversation", dialogStateProperty);
            DialogSet dialogSet = new(accessor);
            dialogSet.Add(this._oauthPrompt);
            return await dialogSet.CreateContextAsync(context, cancellationToken);
        }

        private async Task<Attachment> CreateOAuthCard(ITurnContext context, CancellationToken cancellationToken = default)
        {
            SignInResource signInResource = await UserTokenClientWrapper.GetSignInResourceAsync(context, this._oauthSettings.ConnectionName, cancellationToken);
            string? link = signInResource.SignInLink;
            TokenExchangeResource? tokenExchangeResource = null;

            if (this._oauthSettings.EnableSso == true)
            {
                tokenExchangeResource = signInResource.TokenExchangeResource;
            }

            return new Attachment
            {
                ContentType = OAuthCard.ContentType,
                Content = new OAuthCard
                {
                    Text = this._oauthSettings.Text,
                    ConnectionName = this._oauthSettings.ConnectionName,
                    Buttons = new[]
                    {
                        new CardAction
                        {
                                Title = this._oauthSettings.Title,
                                Text = this._oauthSettings.Text,
                                Type = "signin",
                                Value = link
                        },
                        },
                    TokenExchangeResource = tokenExchangeResource,
                    TokenPostResource = signInResource.TokenPostResource
                },
            };
        }
    }
}
